packages/office-addin-manifest/src/manifestInfo.ts

```diff
@@ -66,110 +65,24 @@ export namespace OfficeAddinManifest {
 
   export async function readManifestFile(manifestPath: string): Promise<ManifestInfo> {
     if (manifestPath) {
-      const xml = await readXmlFromManifestFile(manifestPath);
-      const manifest: ManifestInfo = parseManifest(xml);
+      const manifestHandler: ManifestHandler = getManifestHandler(manifestPath);
+      const fileData: string = await readFromManifestFile(manifestPath);
+
+      const manifest: ManifestInfo = await manifestHandler.parseManifest(manifestPath, fileData);
       return manifest;
     } else {
       throw new Error(`Please provide the path to the manifest file.`);
     }
   }
 }
 
-function parseManifest(xml: Xml): ManifestInfo {
-  const manifest: ManifestInfo = new ManifestInfo();
-  const officeApp: Xml = xml.OfficeApp;
-
-  if (officeApp) {
-    const defaultSettingsXml: Xml = xmlMethods.getXmlElement(officeApp, "DefaultSettings");
-
-    manifest.id = xmlMethods.getXmlElementValue(officeApp, "Id");
-    manifest.allowSnapshot = xmlMethods.getXmlElementValue(officeApp, "AllowSnapshot");
-    manifest.alternateId = xmlMethods.getXmlElementValue(officeApp, "AlternateId");
-    manifest.appDomains = xmlMethods.getXmlElementsValue(officeApp, "AppDomains", "AppDomain");
-    manifest.defaultLocale = xmlMethods.getXmlElementValue(officeApp, "DefaultLocale");
-    manifest.description = xmlMethods.getXmlElementAttributeValue(officeApp, "Description");
-    manifest.displayName = xmlMethods.getXmlElementAttributeValue(officeApp, "DisplayName");
-    manifest.highResolutionIconUrl = xmlMethods.getXmlElementAttributeValue(officeApp, "HighResolutionIconUrl");
-    manifest.hosts = xmlMethods.getXmlElementsAttributeValue(officeApp, "Hosts", "Host", "Name");
-    manifest.iconUrl = xmlMethods.getXmlElementAttributeValue(officeApp, "IconUrl");
-    manifest.officeAppType = xmlMethods.getXmlAttributeValue(officeApp, "xsi:type");
-    manifest.permissions = xmlMethods.getXmlElementValue(officeApp, "Permissions");
-    manifest.providerName = xmlMethods.getXmlElementValue(officeApp, "ProviderName");
-    manifest.supportUrl = xmlMethods.getXmlElementAttributeValue(officeApp, "SupportUrl");
-    manifest.version = xmlMethods.getXmlElementValue(officeApp, "Version");
-
-    if (defaultSettingsXml) {
-      const defaultSettings: DefaultSettings = new DefaultSettings();
-
-      defaultSettings.requestedHeight = xmlMethods.getXmlElementValue(defaultSettingsXml, "RequestedHeight");
-      defaultSettings.requestedWidth = xmlMethods.getXmlElementValue(defaultSettingsXml, "RequestedWidth");
-      defaultSettings.sourceLocation = xmlMethods.getXmlElementAttributeValue(defaultSettingsXml, "SourceLocation");
-
-      manifest.defaultSettings = defaultSettings;
-    }
-  }
-
-  return manifest;
-}
-
-async function modifyManifestXml(manifestPath: string, guid?: string, displayName?: string): Promise<Xml> {
+async function readFromManifestFile(manifestPath: string): Promise<string> {
```

Comment:
Seems like this should be part of the handlers.  It looks like previously this was done as part of the readXmlFromManifestFile and now you've split reading from parsing, but is there any reason you would read w/out parsing or need to parse something that was already read?  You are right in that we don't want to read a file multiple times, but I'm not sure it's the case that we would unless explicitly called to do (or at least we could make it so).

Reply:
This separation was made because now we need to check whether a file is JSON or not before assigning a handler.
To verify that, we need to read the file. That's why this `readFromManifestFile` was extracted from the handler: to be used to check the file extension.
I think it would be nice for the file to only be read once, which is currently the case with this implementation.

Reply:
You don't need to read the file to check the extension . . . you are already doing that and picking the handler before reading.

Was the code before your change reading it more than once?  How does this prevent it from being read more than once?

Reply:
Nah, you are right. I was thinking of a previous implementation I was making.
I added the file reading part to the `ManifestHandler` class, so the code is cleaner.